Add Error Prone code style verification

vlsi opened the pull request:
See https://errorprone.info/

It allows to capture errors like mising switch case branches,
missing override, non-static inner class, etc

The verification can be run locally via ./gradlew -PenableErrorprone classes
Note: Java 11+ is required

The changes do not seem to be very big for me, however, it seems to flag important issues: it forbids the use of deprecated methods, it flags the use of the default encoding, it suggests to avoid `LinkedList`, and so on.

I fixed the warnings by running `classes -PenableErrorprone -Pwerror=false --rerun-tasks` in my IDE, so it printed all the warnings, then I could click on each error to navigate and resolve it.

The biggest commits are

91851b17974b1fd2744414ada6b2ac7937e82735 Replace LinkedList with ArrayList
 89 files changed, 249 insertions(+), 241 deletions(-)

d668dd388105b6b8d297c937026980e54fe8764c Remove uses of deprecated APIs
 52 files changed, 154 insertions(+), 90 deletions(-)

be8438d2cfd517dda856d2674bfff096b8a0445a Suppress DefaultCharset usage: either suppress the warning or use a charset
 27 files changed, 85 insertions(+), 94 deletions(-)

vlsi committed:
Add Error Prone code style verification

See https://errorprone.info/

It allows to capture errors like mising switch case branches,
missing override, non-static inner class, etc

The verification can be run locally via ./gradlew -PenableErrorprone classes
Note: Java 11+ is required

vlsi committed:
Remove LoggingManager class (it has been deprecated since JMeter 3.2)

vlsi committed:
Use UTF-8 encoding in BeanShellClient instead of platform-specific one

vlsi committed:
Add JavaDoc summaries to RandomVariableConfig, CounterConfig, JavaSamplerContext, DynamicClassLoader

vlsi committed:
Suppress error-prone errors for junit-sample

JUnitAmbiguousTestClass is intentional there

vlsi committed:
Replace obsolete Vector and Hashtable with ArrayList and HashMap

vlsi committed:
Replace LinkedList with ArrayList

vlsi committed:
Remove uses of deprecated APIs

vlsi committed:
Remove excessive parenthesis

vlsi committed:
Avoid unsafe reflection cast

pmouawad commented on src/bshclient/src/main/java/org/apache/jmeter/util/BeanShellClient.java:
```diff
@@ -80,7 +82,7 @@ public static void main(String [] args) throws Exception{
     throws IOException
     {
-        outPipe.write( line.getBytes() ); // TODO - charset?
+        outPipe.write(line.getBytes(StandardCharsets.UTF_8));
```
Are you sure about the potential breaks this might introduce ?

vlsi replied:
I believe there's a movement to use UTF-8 by default.
* https://openjdk.java.net/jeps/8187041
* Kotlin uses UTF-8 by default

Of course, we can make it configurable, however, I believe, most of the times UTF-8 by default would be better (less chances to corrupt data) than use platform-default encoding.

pmouawad commented on src/jorphan/src/main/java/org/apache/log/ContextMap.java:
```diff
@@ -53,7 +54,7 @@
 
     ///Container to hold map of elements
-    private Hashtable m_map = new Hashtable();
+    private HashMap<String, Object> m_map = new HashMap<>();
```
Are you sure synchronization is not required ?

vlsi replied:
`ContextMap` was marked as `@deprecated Will be dropped in 3.3`.
Do we really need to investigate that?

I might drop the class altogether if you like.

vlsi replied:
On the other hand, I should probably just suppress all the warnings for the class.

pmouawad reviewed (changes requested):
Hello @vlsi ,
Thanks for this PR.
Just few notes.
Regards

vlsi committed:
Avoid use of Enumeration in ClassFinder

vlsi committed:
Fixup JavaDocs

vlsi committed:
Make constructors of abstract classes protected rather than public

vlsi committed:
Avoid import Map.Entry to avoid amboguity

Importing nested classes/static methods/static fields with
commonly-used names can make code harder to read,
because it may not be clear from the context exactly which
type is being referred to

See https://errorprone.info/bugpattern/BadImport

vlsi committed:
Use lower-case field names for non-static fields in StatCalculator

vlsi committed:
Make enums immutable

vlsi committed:
Suppresss MissingSummary and EmptyBlockTag for now

The code contains too many JavaDoc warnings

vlsi committed:
Suppress deprecation warnings for TestResultAction

vlsi committed:
Suppress deprecation warnings in SamplerMetricFixedModeTest

vlsi committed:
Suppress DefaultCharset usage: either suppress the warning or use a charset

vlsi committed:
Fix InconsistentCapitalization warnings

vlsi committed:
Suppress JdkObsolete warnings for Enumeration

In all the cases Enumeration comes from Java interfaces,
so we can't avoid it.

Unfortunately error-prone does not recognize that yet.

See https://github.com/google/error-prone/issues/1646

vlsi committed:
Remove unused variables

vlsi committed:
Suppress JdkObsolete warnings for usages of Date

It would be nice to migrate to java.time API, so the idea is
to suppress the current usages, and use newer APIs for the new code.

vlsi committed:
Add missing @Override annotations

vlsi committed:
Mark empty catch blocks with a comment

error-prone requires the comment

vlsi committed:
Suppress FutureReturnValueIgnored

Most of the times, the returned Future objects must not be ignored.
DiskStoreSampleSender should probably be reworked to propagate exceptions.

vlsi committed:
Suppress Thread.yield usage

vlsi committed:
Suppress JavaTimeDefaultTimeZone

vlsi committed:
Inline format specifiers when they are used only once

See https://errorprone.info/bugpattern/InlineFormatString

vlsi committed:
Resolve MixedMutabilityReturnType: make the returned lists unmodifiable

See https://errorprone.info/bugpattern/MixedMutabilityReturnType

vlsi committed:
Suppress UnnecessaryAnonymousClass as the only two warnings seem to be false positives

See https://errorprone.info/bugpattern/UnnecessaryAnonymousClass

vlsi committed:
Resolve SynchronizeOnNonFinalField

vlsi committed:
Suppress warnings for Hashtable usage when the class is required for Java API

vlsi committed:
Move .lock() statements out of try to avoid accidetal release if lock fails

See https://errorprone.info/bugpattern/LockNotBeforeTry

vlsi committed:
Suppress StaticAssignmentInConstructor in Summarizer

vlsi committed:
Suppress JdkObsolete when using appendReplacement(StringBuffer,...)

Unfortunately, error-prone can't detect that false positive yet

vlsi committed:
Suppress JavaLangClash for ThreadGroup

vlsi committed:
Avoid NarrowingCompoundAssignment warning

int += long pattern hides a lossy cast, so it is better to write that explicitly

vlsi committed:
Make members of final classes package-private

See https://errorprone.info/bugpattern/ProtectedMembersInFinalClass

vlsi commented:
@pmouawad , the PR passes tests, and it looks like it is ready for review and merge.

ham1 commented on extras/startup.bsh:
```diff
@@ -49,7 +48,7 @@ print(p + " = " + getprop(p));
 
 loglevel(String priority, String category){
-LoggingManager.setPriority(priority, category);
+// TODO: implement?
```
Can we do the TODO before merging, or put the current behaviour back?

vlsi replied:
Pull requests are welcome.
I have no idea what this code means, however, I am 100% sure the previous implementation was doing absolutely nothing since `LogginManager.setPriority` was an empty method.

ham1 replied:
Understood. Could we just remove the TODO? Or even better state this has always been unimplemented?

ham1 commented on src/components/src/main/java/org/apache/jmeter/assertions/XMLAssertion.java:
```diff
@@ -45,10 +48,12 @@
         protected XMLReader initialValue() {
             try {
-                XMLReader reader = XMLReaderFactory.createXMLReader();
+                SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
```
Do we need all the intermediary variables? e.g.
```
XMLReader reader =
        SAXParserFactory.newInstance().newSAXParser().getXMLReader();
````

FSchumacher commented:
My biggest concern with this PR at the moment is its size. Could we add the feature (errorprone) without changing every nag it shows? I tend to get bored when looking at big PRs and either skip parts, or abort looking at it altogether.
Apart from that, I am +1 on adding the feature (if it doesn't break the build when running with Java 8). (Haven't tried it)

vlsi commented:
> if it doesn't break the build when running with Java 8

All the previous CI jobs pass, so it works.
In practice, it is activated only in case `-PenableErrorprone` is added as a command-line option or `gradle.properties` setting.

> Could we add the feature (errorprone) without changing every nag it shows? 

Well, most of the warnings are true issues. There is an unfortunate case when the tool discourages the use of `Hashtable`, `Enumeration`, and `LinkedList` in the same check.
Nobody uses `Hashtable` nowadays unless that is required for an ancient API, however, people can easily use `LinkedList` without realizing it is way slower and less efficient than `ArrayList` and `ArrayDeque`
